### ISSUE TITLE
[BUG] Performance degradation 

### ISSUE THREAD

**Issue by mrfelton** [nx-dotnet/nx-dotnet#900]
### Current Behavior

We just updated to the latest version of nx-dotnet which uses `dotnet list reference` to get the list of project dependencies. This has the unfortunate downside of increasing the time it takes to generate the dependency graph.

For one, it adds a dependency on dotnet to run commands like `nx affected` where previously there wasn't this dependency. Every workload or CI pipeline step that needs to call these nx commands now needs to have dotnet installed on the system, and running the commands appear to take considerably longer than the prior native js version.

### Expected Behavior

Performance should continuously improve with new releases, rather than getting worse.

### Steps to Reproduce

1. run `nx affected` with large monorepo using a version of nx-dotnet from before the move to depending on dotnet and compare run times with the new version.

### Nx Report

```shell
n/a
```

### nx.json

```json
n/a
```

**Comment by AgentEnder** [comment]
Its true that the native JS version was faster, but it was also quite incomplete. Projects referenced by glob patterns inside files were missed, references coming from Directory.build.props would not be considered, etc. I don't like it getting slower either, but without getting the deps directly from .NET, its going to be a lot lot harder to have confidence in the fact that your graph is correct.

If your graph is not correct, you can easily end up not running tests on affected code, or not deploying things that need to be deployed etc.

We can't just revert that changeset back for these reasons. If you or anyone else is interested in finding a way to speed up the dotnet call execution, or can find a way to identify deps **_correctly_** without invoking .NET, I'd love to see it.

**Comment by mrfelton** [comment]
I understand the rationale and it makes sense, it's unfortunate that performance has taken a hit in the process.

> references coming from Directory.build.props would not be considered

They are still not considered, unfortunately.

**Comment by AgentEnder** [comment]
> They are still not considered, unfortunately.

Really? That doesn't make sense as if that's the case I'd expect that they also aren't included during dotnet build.

**Comment by Tommy228** [comment]
What about invoking a .NET program that uses the official Microsoft API to read and extract included projects from all projects ([EvaluatedInclude](https://learn.microsoft.com/en-us/dotnet/api/microsoft.build.evaluation.projectitem.evaluatedinclude?view=msbuild-17-netcore) This could be significantly faster, but then a separate .NET app would have to be maintained.

**Comment by AgentEnder** [comment]
I don't think this would be faster, you'd have to invoke the built all with the dotnet CLI still, and I imagine the .NET team would be better at maintaining the CLI command then anything we shipped

**Comment by mrfelton** [comment]
>> They are still not considered, unfortunately.
>
>Really? That doesn't make sense as if that's the case I'd expect that they also aren't included during dotnet build.

Yes, see example here: https://github.com/mrfelton/nx-dotnet-depgraph-issue-reproducer

It seems that `dotnet list reference` does not consider dependencies added through `Directory.Build.props`

**Comment by mrfelton** [comment]
Some examples of other standalone projects that attempts to generate dependency information.

- https://github.com/bjorkstromm/depends
- https://github.com/phmonte/Buildalyzer
- https://github.com/dswisher/dotnet-list-refs

**Comment by bcallaghan-et** [comment]
> > They are still not considered, unfortunately.
> 
> Really? That doesn't make sense as if that's the case I'd expect that they also aren't included during dotnet build.

I've downloaded @mrfelton's demo repo, and I can confirm his results. The `dotnet build` command properly recognizes that `my-third-lib` is a dependency of the root `my-lib`, but the `dotnet list reference` command doesn't include it. While that may be a bug in .NET itself, it does mean the Nx graph is inaccurate in that scenario.

I'm also seeing the performance slow down in my .NET monorepo (200+ projects). I was testing Nx as a way to speed up our CI builds, but the slow read of `dotnet list reference` almost entirely negates the performance gains from Nx itself. At this point, I may need to downgrade to an older version of nx-dotnet that examines the csproj files directly. My repository isn't doing anything unusual with references, so the old approach still works.
